### DISCUSSION TITLE
Moving forward with new packaging for UI framework features

### DISCUSSION THREAD

**Discussion by idg10** [dotnet/reactive#2177] [Announcements]
We are moving forward with our plans to modify how UI-framework-specific features are packaged in Rx.NET. This is the last chance to offer feedback on #2086

Back in November 2023, we announced that we intended to fix some long-standing problems by making [changes to Rx.NET packaging](https://github.com/dotnet/reactive/discussions/2038).

In February 2024, we created PR #2086 which enumerated the various options under consideration (and described all the suggestions we'd had from the community since the initial announcement).

Also in February we created PR #2084 as a prototype showing how this change might look, with some changes in March in response to community feedback.

.NET 9.0 is due to ship very soon. We intend to produce a new Rx.NET release not long after that happens.

This new version seems like the right time to move forward with the packaging plan. The first phase is to make new UI-framework-specific Rx packages available. We will not remove anything from `System.Reactive` at this time, but will mark all UI-framework-specific types and methods as `[Obsolete]`, beginning the process 

(Note: Rx 6.0 will work on .NET 9.0—anyone using Rx 6.0 today won't need to upgrade just to be able to move to .NET 9.0. The motivation for a new Rx.NET release is mainly that .NET 6.0 will be going out of support. And a secondary motivation is that a lot of people don't really understand how TFMs work, and mistakenly think that the absence of a `net8.0` target indicates a lack of support for .NET 8.0. In fact we've supported Rx 6.0 on .NET 8.0 since .NET 8.0 shipped. But we want to make it clear that .NET 6.0 will not be officially supported by Rx.NET once it is no longer supported by Microsoft. So Rx 7.0 will not offer `net6.0` TFMs. We plan to offer `netstandard2.0`, `net472`, `net8.0`, `net8.0-windows10.0.19041` and `uap10.0.18632`.)

**Comment by Sergio0694** [comment]
This sounds awesome! I have a couple questions:
- Would this first phase also include the new "base" package mentioned in https://github.com/dotnet/reactive/discussions/2038#discussioncomment-7557014?
- This sounds like the perfect occasion to also add .NET 9 support for the UWP flavor of `System.Reactive` (see [the announcement](https://devblogs.microsoft.com/ifdef-windows/preview-uwp-support-for-dotnet-9-native-aot/)). Would this be something you'd accept as a contribution? Without this, there would be no way for UWP apps to continue using `System.Reactive` when moving to .NET 9 (eg. that applies to us too in the Microsoft Store app) . I'd love to help with this🙂

**Comment by idg10** [comment]
> Would this first phase also include the new "base" package

This one encountered significant push back from the community, and so we have instead decided to go down the slower, but less disruptive path of keeping `System.Reactive` as the main package. Since we found a workaround for the package bloat problems that were being caused by having the UI frameworks integrated into `System.Reactive`, it was no longer an absolute requirement to separate them out at the earliest opportunity.

> .NET 9 support for [...] UWP

I was briefly excited when I saw that announcement, but as I mentioned in the UAP ADR (currently on the .NET 9.0 SDK branch), unfortunately it seems to impose a minimum target version of Windows 11 22H2 (26100). Windows 10 will remain in support for a few years yet, so we can't require that.

**Reply by Sergio0694** [reply]
> "unfortunately it seems to impose a minimum target version of Windows 11 22H2 (26100)"

It doesn't! It requires 26100 as **target** version, not as minimum. The minimum can be as low as you want (eg. 17763). Additionally, that requirement **only** applies to projects using XAML or PRI resources, and `System.Reactive` doesn't do either of those things (ie. it doesn't need the XAML compiler nor the APPX tools at all). So we can also lower the TFM if we wanted. But even if we kept 26100 as target, that would be completely fine, as that'd just be its own TFM in the `.Uwp` flavor of the Reactive NuGet package, along with `uap10.0`. Only developers on UWP on .NET 9 would be impacted by that, which is also fine. That is, you can keep the legacy UWP binaries (ie. `uap10.0`) using a lower TFM, and only bump that for UWP on .NET 9. The two are entirely separate anyway (different TFMs). So you'd be able to keep supporting Windows 10 for existing UWP users no matter what.

I'm happy to chat more about this if you need more context 🙂
But basically the TLDR is this should not be a problem at all for Reactive, I believe.

**Reply by kmgallahan** [reply]
@idg10 

>  Since we found a workaround for the package bloat problems that were being caused by having the UI frameworks integrated into System.Reactive, it was no longer an absolute requirement to separate them out at the earliest opportunity.

Unless I'm mistaken or we're talking about different things, this doesn't work for WinUI 3 projects.

See my last 2 comments on this thread: https://github.com/dotnet/reactive/discussions/2038#discussioncomment-8546452

**Reply by Sergio0694** [reply]
Mmh I don't see how that workaround could work. Wouldn't that disable **all** transitive framework references? But you'd still need some of them for UWP/WinUI apps (eg. the `Microsoft.Windows.SDK.NET.Ref` one you mentioned). Makes sense to me that when you tried testing this in a WinUI app, things broke. I don't think this scenario is even supported at all?

**Reply by kmgallahan** [reply]
@Sergio0694 

It might work if a change is made to detect `UseWinUI` as was done for WinForms and WPF here: https://github.com/dotnet/reactive/pull/859

**Reply by Sergio0694** [reply]
Not sure I understand how that works. Even if they had a .targets that manually removed the reference on the System.Reactive .dll from the -windows TFM and used the one from the non-Windows one, the NuGet package would still declare a transitive framework reference for WindowsDesktop.App, which means you'd still have the .NET SDK pull a bunch of unwanted stuff in. And even doing more tricks to make to make it work, this is still ultimately fighting against the .NET SDK and NuGet, and breaking pretty much every recommended standard and convention on how to structure a NuGet package.

Personally I don't see a proper way to fix things without just having a separate, vanilla .NET 8, "Core" package, which the other UI-specific NuGet packages could then just pull in as a transitive dependency, and add extensions for depending on each UI framework.

**Reply by Sergio0694** [reply]
I actually found another way to solve the issue entirely for us:
- Use `PackageDownload` to reference System.Reactive
- Manually add a `Reference` item for the `net6.0` .dll

You get full System.Reactive, with no WindowsForms/WPF references whatsoever 😄
I'll just do this in the meantime, works just fine and completely removes all unwanted cruft.

**Reply by kmgallahan** [reply]
@Sergio0694

You mean a ref to `netstandard.dll` in `...dotnet\sdk...`?

Would this work when a 3rd party package has a ref to System.Reactive (e.g. DynamicData)?

**Reply by Sergio0694** [reply]
No, I meant a reference to `\lib\net6.0\System.Reactive.dll` in the NuGet package 🙂

**Reply by kmgallahan** [reply]
In a .Net lib that refs System.Reactive & DynamicData, after attempting to use `PackageDownload` & `<Reference>` with `<HintPath>` to point to the dlls I'm getting I'm getting

```
... error CS0012: The type 'IAsyncOperation<>' is defined in an assembly that is not referenced. You must add a reference to assembly 'Microsoft.Windows.SDK.NET, Version=10.0.19041.24, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
... error CS0012: The type 'IAsyncOperationWithProgress<,>' is defined in an assembly that is not referenced. You must add a reference to assembly 'Microsoft.Windows.SDK.NET, Version=10.0.19041.24, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
```

**Reply by kmgallahan** [reply]
Ehhh, resolved the `IAsyncOperation` issues, but now I have the same errors as before when disabling transitive refs:

```
error CS7069: Reference to type 'Dispatcher' claims it is defined in 'WindowsBase', but it could not be found
error CS7069: Reference to type 'DispatcherObject' claims it is defined in 'WindowsBase', but it could not be found
error CS0012: The type 'Control' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.Windows.Forms, Version=6.0.2.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.
```

Seems the code for these things is in the platform files, which have references to the old framework namespaces.

https://github.com/dotnet/reactive/tree/main/Rx.NET/Source/src/System.Reactive/Platforms

https://github.com/dotnet/reactive/blob/main/Rx.NET/Source/src/System.Reactive/Platforms/Desktop/Linq/ControlObservable.cs

**Reply by Sergio0694** [reply]
Yeah, that's why I specifically added a reference to the `net6.0` .dll, **not** to the `net6.0-windows10.0.19041` one 🙂

**Reply by kmgallahan** [reply]
Still have a pile of framework files ¯\\\_(ツ)_/¯

![Screenshot 2024-11-11 174945](https://github.com/user-attachments/assets/8bc02295-4f25-4c2a-8186-ef4ce7a7c4fa)

**Reply by Sergio0694** [reply]
Not sure how that's possible assuming you're also using `PackageDownload`. You'd need to get a binlog to figure out where they're coming from. I've tested in our repo and publishing with Native AOT did result in no WPF references left with this setup.

**Reply by kmgallahan** [reply]
I forgot to respond here. If I remember correctly, the process of elimination indicated that the FlaUI package was responsible for the unwanted files due to its dependencies.

**Reply by idg10** [reply]
OK, in the light of this discussion, we did _not_ move forward as we thought we would.

There is now a new plan: https://github.com/dotnet/reactive/discussions/2211
